### PR TITLE
✨ Add torchfx feature extractor

### PR DESCRIPTION
**Pull request by ashwinvaidya17**
# Description

This PR only introduces torchfx. For other refactor see #717 

- Add torchfx feature extractor
- Rename existing one to timm
- Add custom backbone support in TorchFX feature extractor

## Changes

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the [pre-commit style and check guidelines](https://openvinotoolkit.github.io/anomalib/guides/using_pre_commit.html#pre-commit-hooks) of this project.
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes

### CODE REVIEW

File: anomalib/models/components/feature_extractors/torchfx.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+"""Feature Extractor based on TrochFX."""
```
Comment (djdameln) [inline anomalib/models/components/feature_extractors/torchfx.py:1]: typo in TorchFX

File: anomalib/models/components/feature_extractors/torchfx.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    Example:
+        >>> import torch
+        >>> from anomalib.models.components.feature_extractors import get_torchfx_feature_extractor
+        >>> from torchvision.models.efficientnet import EfficientNet_B5_Weights
+        >>> feature_extractor = get_torchfx_feature_extractor(
+                backbone="efficientnet_b5", return_nodes=["features.6.8"], weights=EfficientNet_B5_Weights.DEFAULT
+            )
+        >>> input = torch.rand((32, 3, 256, 256))
+        >>> features = feature_extractor(input)
+        >>> [layer for layer in features.keys()]
```
Comment (ashwinvaidya17) [inline anomalib/models/components/feature_extractors/torchfx.py:35]: Should say `features.6.8`

File: anomalib/models/padim/torch_model.py [hunk 3]
```diff
@@ -42,7 +42,7 @@ def __init__(
 
         self.backbone = backbone
         self.layers = layers
-        self.feature_extractor = FeatureExtractor(backbone=self.backbone, layers=layers, pre_trained=pre_trained)
+        self.feature_extractor = TimmFeatureExtractor(backbone=self.backbone, layers=layers, pre_trained=pre_trained)
```
Comment (djdameln) [inline anomalib/models/padim/torch_model.py:45]: Why don't we switch to the torch fx feature extractor in our models?

File: anomalib/models/components/feature_extractors/torchfx.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+import importlib
+from typing import List, Optional
+
+from torch.fx.graph_module import GraphModule
+from torchvision.models._api import WeightsEnum
+from torchvision.models.feature_extraction import create_feature_extractor
+
+
+def get_torchfx_feature_extractor(
```
Comment (djdameln) [inline anomalib/models/components/feature_extractors/torchfx.py:14]: Do we really need this as a separate function? We could just call `create_feature_extractor` directly from the code right?

Comment (ashwinvaidya17) [reply anomalib/models/components/feature_extractors/torchfx.py:14]: We also have to freeze the gradients. We might miss doing this in the models. This provides an interface similar to our existing implementation. But instead of `pretrained` we have `weights` now as pretrained is going to be deprecated in torchvision.

File: anomalib/models/padim/torch_model.py (hunk 3 above)
Comment (ashwinvaidya17) [reply anomalib/models/padim/torch_model.py:45]: I don't mind doing that.

File: anomalib/models/components/feature_extractors/torchfx.py (hunk 4 above)
Comment (djdameln) [reply anomalib/models/components/feature_extractors/torchfx.py:14]: I see. In that case this is a good solution. I'm surprised torchvision does not freeze the gradients by default in `create_feature_extractor`, if it reduces memory use.

File: anomalib/models/components/feature_extractors/torchfx.py (hunk 1 above)
Comment (samet-akcay) [inline anomalib/models/components/feature_extractors/torchfx.py:1]:
```suggestion
"""Feature Extractor based on TorchFX."""
```

File: anomalib/models/components/feature_extractors/torchfx.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    backbone: str, return_nodes: List[str], weights: Optional[WeightsEnum] = None
+) -> GraphModule:
+    """Extract features from a CNN.
+
+    Args:
+        backbone (nn.Module): The backbone to which the feature extraction hooks are attached.
+        return_nodes (Iterable[str]): List of layer names of the backbone to which the hooks are attached.
+            You can find the names of these nodes by using `get_graph_node_names` function.
+        weights (Optional[WeightsEnum]): Weights enum to use for the model.
+            These enums are defined in `torchvision.models.<model>`.
```
Comment (samet-akcay) [inline anomalib/models/components/feature_extractors/torchfx.py:24]:
```suggestion
            These enums are defined in ``torchvision.models.<model>``.
```

File: anomalib/models/cflow/torch_model.py [hunk 6]
```diff
@@ -38,7 +38,7 @@ def __init__(
         self.dec_arch = decoder
         self.pool_layers = layers
 
-        self.encoder = FeatureExtractor(backbone=self.backbone, layers=self.pool_layers, pre_trained=pre_trained)
+        self.encoder = TimmFeatureExtractor(backbone=self.backbone, layers=self.pool_layers, pre_trained=pre_trained)
```
[the remark this answers is not in the document]
Comment (ashwinvaidya17) [reply anomalib/models/cflow/torch_model.py:41]: We can make it configurable. Maybe we can mention the discrepancy in a comment but I think Timm feature extractor would be necessary in OTX where we have constraints on the torchvision version

File: anomalib/models/padim/torch_model.py (hunk 3 above)
Comment (ashwinvaidya17) [reply anomalib/models/padim/torch_model.py:45]: Let's keep it for now till otx stuff is sorted.

File: tests/pre_merge/models/test_feature_extractor.py [hunk 7]
```diff
@@ -1,7 +1,11 @@
 import pytest
 import torch
+from torchvision.models.efficientnet import EfficientNet_B5_Weights
 
-from anomalib.models.components.feature_extractors import FeatureExtractor
+from anomalib.models.components.feature_extractors import (
+    TimmFeatureExtractor,
+    get_torchfx_feature_extractor,
```
Comment (samet-akcay) [inline tests/pre_merge/models/test_feature_extractor.py:7]: Would it be an idea to have `TorchFeatureExtractor` just for the naming consistency with `TimmFeatureExtractor`? It would use `get_torchfx_feature_extractor` in its `__call__`. Or would it be an overkill?

File: anomalib/models/components/feature_extractors/timm.py [hunk 8]
```diff
@@ -23,9 +23,9 @@ class FeatureExtractor(nn.Module):
 
     Example:
         >>> import torch
-        >>> from anomalib.core.model.feature_extractor import FeatureExtractor
+        >>> from anomalib.models.components.feature_extractors import TimmFeatureExtractor
 
-        >>> model = FeatureExtractor(model="resnet18", layers=['layer1', 'layer2', 'layer3'])
+        >>> model = TimmFeatureExtractor(model="resnet18", layers=['layer1', 'layer2', 'layer3'])
```
Comment (jpcbertoldo) [inline anomalib/models/components/feature_extractors/timm.py:28]: I didn't duly read this PR but i saw this and wanted just to point a general suggestion.

I think that wherever this type of documentation examples include imports and/or anomalib classes the docstring should just point to one of those `.ipynb` tutorials/examples (instead of having a functional example in the docstring).

I think this would be less painful to update as the package evolves quite fast -- it occured to me more than twice that the example I got was not actually function.

**Comment by samet-akcay** [comment]
Another comment would be that this PR would break the backward compatibility due to the removal of `FeatureExtractor`. It would be good to still support it such that when the user tries to import `from anomalib.models.components import FeatureExtractor` it would inherit `TimmFeatureExtractor` but throws a deprecation warning.

File: anomalib/models/components/feature_extractors/timm.py (hunk 8 above)
Comment (ashwinvaidya17) [reply anomalib/models/components/feature_extractors/timm.py:28]: Maybe it would be a good idea to have both. Personally, I like to see the usage in hints and it also makes autodoc more descriptive. While ipynb files are good for getting started, I feel they introduce an extra overhead for maintenance. If there are broken examples then that's something to keep in mind on our part and future PR reviews. Examples make the method more readable so I am in favour of keeping these.

**Comment by ashwinvaidya17** [comment]
> 

Good point

File: anomalib/models/components/feature_extractors/torchfx.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+        Returns:
+            Feature Extractor based on TorchFX.
+        """
+        backbone_model = self.get_backbone_class(backbone.class_path)
+        if isinstance(weights, WeightsEnum):  # torchvision models
+            feature_extractor = create_feature_extractor(
+                backbone_model(weights=weights, **backbone.init_args), return_nodes
+            )
+        else:
+            backbone_model = backbone_model(**backbone.init_args)
```
Comment (samet-akcay) [inline anomalib/models/components/feature_extractors/torchfx.py:114]:
```suggestion
            backbone_model = backbone_class(**backbone.init_args)
```

File: anomalib/models/components/feature_extractors/torchfx.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+                models = importlib.import_module("torchvision.models")
+                backbone_model = getattr(models, backbone)
+        except ModuleNotFoundError as exception:
+            raise ModuleNotFoundError(
+                f"Backbone {backbone} not found in torchvision.models and not found in {backbone} module either"
+            ) from exception
+
+        return backbone_model
+
+    def __call__(self, inputs: Tensor) -> Dict[str, Tensor]:
```
Comment (samet-akcay) [inline anomalib/models/components/feature_extractors/torchfx.py:153]: It would be an idea to decide on the name of the variable. We use `input`, `inputs` and `input_tensor` interchangeably. As far as I know, PyTorch prefers `input`. It might be an idea to settle down on a single variable name

**Review by samet-akcay — CHANGES_REQUESTED** [review]
Thanks @ashwinvaidya17 for your great effort! Only minor comments

File: anomalib/models/components/feature_extractors/torchfx.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+    def __init__(
+        self,
+        backbone: Union[str, BackboneParams, Dict],
+        return_nodes: List[str],
+        weights: Optional[Union[WeightsEnum, str]] = None,
+        requires_grad: bool = False,
+    ):
+        if isinstance(backbone, dict):
+            backbone = BackboneParams(**backbone)
+        elif isinstance(backbone, str):
```
[the remark this answers is not in the document]
Comment (ashwinvaidya17) [reply anomalib/models/components/feature_extractors/torchfx.py:79]: The default case is to assume that backbone is of type `BackboneParams`. I am using `elif` just to reduce one check if the first one succeeds.

File: anomalib/models/components/feature_extractors/torchfx.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+            feature_extractor.eval()
+            for param in feature_extractor.parameters():
+                param.requires_grad_(False)
+
+        return feature_extractor
+
+    def get_backbone_class(self, backbone: str) -> nn.Module:
+        """Get the backbone class from the provided path.
+
+        If only the moodel name is provided, it will try to load the model from torchvision.
```
[the remark this answers is not in the document]
Comment (ashwinvaidya17) [reply anomalib/models/components/feature_extractors/torchfx.py:130]: Yeah example would be a good idea

File: anomalib/models/components/feature_extractors/torchfx.py (hunk 10 above)
Comment (ashwinvaidya17) [reply anomalib/models/components/feature_extractors/torchfx.py:153]: How about `inputs` as we type it as `Tensor`. `input` might clash with built in `input` method.

**Review by samet-akcay — APPROVED** [review]
Thanks for addressing my comments

File: tests/pre_merge/models/test_feature_extractor.py [hunk 13]
```diff
@@ -39,6 +42,32 @@ def test_feature_extraction(self, backbone, pretrained):
         else:
             pass
 
+    def test_torchfx_feature_extraction(self):
```
Comment (djdameln) [inline tests/pre_merge/models/test_feature_extractor.py:45]: Maybe add a test case for creating a feature extractor from a local model definition, and a string to a weights file. That way we would have a more complete coverage of the possible use cases of the class.

We could use a dummy model consisting of a few convolutional layers and corresponding weights file.

**Review by djdameln — COMMENTED** [review]
Thanks, just one suggestion for improving the unit tests.

File: tests/pre_merge/models/test_feature_extractor.py (hunk 13 above)
Comment (ashwinvaidya17) [reply tests/pre_merge/models/test_feature_extractor.py:47]: I've added functionality to directly pass locally defined class + tests for it

**Review by djdameln — APPROVED** [review]
Thanks!

**Comment by jpcbertoldo** [comment]
sorry @ashwinvaidya17 i couldn't find the time to review this one
